_posts/2021-05-07-spring-cloud-aws-rds.md

```diff
@@ -0,0 +1,363 @@
+
+## Configuring the Read-Replica for Increasing Throughput
+Replication is a process by which we can copy data from one database server (also known as source database) to be copied to one or more database servers (known as replicas). It is a feature of the database engines of MariaDB, Microsoft SQL Server, MySQL, Oracle, and PostgreSQL DB which can be configured with AWS RDS. 
+
+Amazon RDS uses this built-in replication feature of these databases to create a special type of DB instance called a read replica from a source DB instance. 
+
+The source DB instance plays the role of the primary DB instance and updates made to the primary DB instance are asynchronously copied to the read replica. 
+
+This way we can increase the overall throughput of the database by reducing the load on our primary DB instance by routing read queries from your applications to the read replica.
+
+Spring Cloud AWS supports the use of read-replicas with the help of Spring Framework's declarative transaction support with read-only transactions. We do this by enabling read-replica support in our data source configuration. 
+
+When read-replica enabled, any read-only transaction will be routed to a read-replica instance and the primary database will be used only for write operations.
+
+We enable read-replica support by setting a property `readReplicaSupport`. Our `application.properties` with this property set looks like this:
+
+```properties
+cloud.aws.credentials.profile-name=pratikpoc
+cloud.aws.region.auto=false
+cloud.aws.region.static=us-east-1
+
+cloud.aws.rds.instances[0].db-instance-identifier=testinstance
+cloud.aws.rds.instances[0].username=pocadmin
+cloud.aws.rds.instances[0].password=pocadmin
+cloud.aws.rds.instances[0].databaseName=mysql
+
+cloud.aws.rds.instances[0].readReplicaSupport=true
+```
+Here we have set the `readReplicaSupport` to true to enable read-replica support.
+
+Our service class with a read-only method looks like this:
+
+```java
+@Service
+public class SystemRepository {
+  
+   private final JdbcTemplate jdbcTemplate;
+
+   @Autowired
+   public SystemRepository(DataSource dataSource) {
+    this.jdbcTemplate = new JdbcTemplate(dataSource);
+   }
+  
+  
+  @Transactional(readOnly = true)
+  public List<String> getUsers(){
+    List<String> result = jdbcTemplate.query("SELECT USER() FROM DUAL", new RowMapper<String>(){
+
+      @Override
+      public String mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return rs.getString(1);
+      }
+      
+    });
+    return result;     
+  }
+
+}
+
+```
+Here we have decorated the method `getUsers` with `Transactional(readOnly = true)`. At runtime, all the invocations of this method will be sent to the read-replica.
```

Comment:
But where did we define the read replica? There is only one data source in the configuration.